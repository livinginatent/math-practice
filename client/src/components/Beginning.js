import { Typography } from "@mui/material";
import React, { useRef, useState } from "react";
import { useEffect } from "react";
import { useDispatch } from "react-redux";
import { startGame, start } from "../features/gameSlice";

function Beginning() {
  const [count, setCount] = useState(3);
  const [message, setMessage] = useState("");

  const dispatch = useDispatch();
  const countRef = useRef(count);

  useEffect(() => {
    countRef.current = count;
  }, [count]);

  const handleCount = () => {
    if (countRef.current === 1) {
      return setMessage("GO");
    }

    return setCount((count) => count - 1);
  };

  useEffect(() => {
    const interval = setInterval(() => {
      handleCount();
    }, 1000);

    if (message === "GO") {
      setTimeout(() => {
        dispatch(startGame());
        dispatch(start());
      }, 1000);
    }
    return () => clearInterval(interval);
  }, [count, message]);

  return (
    <>
      <Typography variant="h1" fontStyle={"Poppins"} fontSize={36}>
        GET READY...
      </Typography>

      <Typography fontSize={48}>{count}</Typography>
      <Typography fontSize={60}>
        <span data-testid="go-message">{message}</span>
      </Typography>
    </>
  );
}

export default Beginning;
