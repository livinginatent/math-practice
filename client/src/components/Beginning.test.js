import renderer from 'react-test-renderer';
import Beginning from './Beginning';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import { render, waitFor } from '@testing-library/react';
import { screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

describe('Testing ‘Beginning’ component', () => {
  const initialState = {
    points: 0,
    lives: 3,
    seconds: 20,
    newGame: false,
    isStarted: false,
    isFinished: false,
  };

  const mockStore = configureMockStore();
  const store = mockStore(initialState);
  const { component } = renderer.create(
    <Provider store={store}>
      <Beginning />
    </Provider>
  );

  test('Matches the snapshot', () => {
    expect(component).toMatchSnapshot();
  });
  test('‘GET READY’ message renders', () => {
    render(
      <Provider store={store}>
        <Beginning />
      </Provider>
    );
    const message = screen.getByText(/GET READY.../i);
    expect(message.innerHTML).toBe('GET READY...');
  });

  test('The numbers should be counting down and display `GO`', async () => {
    jest.useFakeTimers();
    render(
      <Provider store={store}>
        <Beginning />
      </Provider>
    );

    const countingDown = screen.getByText('3');
    const message = screen.getByTestId('go-message');

    jest.advanceTimersByTime(1000);
    await waitFor(() => {
      expect(Number(countingDown.innerHTML)).toBe(2);
    });
    jest.advanceTimersByTime(1000);
    await waitFor(() => {
      expect(Number(countingDown.innerHTML)).toBe(1);
    });
    jest.advanceTimersByTime(1000);
    await waitFor(() => {
      expect(message.innerHTML).toBe('GO');
    });
  });

  test('should dispatch with start()', async () => {
    render(
      <Provider store={store}>
        <Beginning />
      </Provider>
    );

    jest.advanceTimersByTime(1000);
    expect(store.getActions()).toEqual([{ type: 'game/start' }]);
  });

  jest.useRealTimers();
});