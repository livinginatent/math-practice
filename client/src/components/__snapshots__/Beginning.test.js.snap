// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ‘Beginning’ component Matches the snapshot 1`] = `undefined`;
